go/test/endtoend/recovery/pitr/pitr_test.go

```diff
@@ -0,0 +1,141 @@
+package pitr
+
+import (
+	"fmt"
+	"testing"
+	"time"
+
+	"github.com/magiconair/properties/assert"
+
+	"vitess.io/vitess/go/vt/log"
+
+	"vitess.io/vitess/go/test/endtoend/cluster"
+
+	"github.com/stretchr/testify/require"
+)
+
+var (
+	createTable = `create table product (id bigint(20) primary key, name char(10), created bigint(20));`
+	insertTable = `insert into product (id, name, created) values(%d, '%s', unix_timestamp());`
+	selectMaxID = `select max(id) from product`
+)
+
+func TestPointInTimeRecovery(t *testing.T) {
+	// create table and insert 2 rows
+	_, err := masterTablet.VttabletProcess.QueryTablet(createTable, keyspaceName, true)
+	require.NoError(t, err)
+	insertRow(t, 1, "p1", false)
+	insertRow(t, 2, "p2", false)
+
+	//start the binlog server and point it to master
+	bs, err := newBinlogServer(hostname, clusterInstance.GetAndReservePort())
+	defer bs.stop()
+	require.NoError(t, err)
+
+	err = bs.start(mysqlMaster{
+		hostname: binlogHost,
+		port:     masterTablet.MysqlctlProcess.MySQLPort,
+		username: mysqlUserName,
+	})
+	require.NoError(t, err)
+
+	// take the backup (to simulate the regular backup)
+	err = clusterInstance.VtctlclientProcess.ExecuteCommand("Backup", replicaTablet.Alias)
+	require.NoError(t, err)
+
+	backups, err := clusterInstance.ListBackups(shardKsName)
+	require.NoError(t, err)
+	require.Equal(t, len(backups), 1)
+
+	// now insert some more data to simulate the changes after regular backup
+	// every insert has some time lag/difference to simulate the time gap between rows
+	// and when we recover to certain time, this time gap will be able to identify the exact eligible row
+	var timeToRecover string
+	for counter := 3; counter <= 7; counter++ {
+		if counter == 5 { // we want to recovery till this, so noting the time
+			tm := time.Now().Add(1 * time.Second).UTC()
+			timeToRecover = tm.Format(time.RFC3339)
+		}
+		insertRow(t, counter, fmt.Sprintf("prd-%d", counter), true)
+
+	}
+
+	// start the recovery
+	recoveryTablet := clusterInstance.NewVttabletInstance("replica", 0, cell)
+	launchRecoveryTablet(t, recoveryTablet, bs, timeToRecover)
+
+	sqlRes, err := recoveryTablet.VttabletProcess.QueryTablet(selectMaxID, keyspaceName, true)
+	require.NoError(t, err)
+
+	assert.Equal(t, sqlRes.Rows[0][0].String(), "INT64(5)")
+	defer recoveryTablet.MysqlctlProcess.Stop()
+	defer recoveryTablet.VttabletProcess.TearDown()
+}
+
+func insertRow(t *testing.T, id int, productName string, isSlow bool) {
+	_, err := masterTablet.VttabletProcess.QueryTablet(fmt.Sprintf(insertTable, id, productName), keyspaceName, true)
+	require.NoError(t, err)
+	if isSlow {
+		time.Sleep(3 * time.Second)
```

Comment:
does this have to be 3 seconds? can it be 1 second? As it is the endtoend tests are quite time-consuming.